source/mesh_refinement/isosurfaces.cc

```diff
@@ -0,0 +1,391 @@
+                return maximum_refinement_level;
+              }
+          }
+        else
+          {
+            return Utilities::string_to_int(string);
+          }
+
+      }
+    }
+
+    template <int dim>
+    void
+    Isosurfaces<dim>::update ()
+    { }
+
+    template <int dim>
+    void
+    Isosurfaces<dim>::tag_additional_cells () const
+    {
+      if (this->get_dof_handler().n_locally_owned_dofs() == 0)
+        return;
+
+      const Quadrature<dim> quadrature(this->get_fe().base_element(this->introspection().base_elements.temperature).get_unit_support_points());
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature,
+                               update_quadrature_points | update_values | update_gradients);
+
+
+      MaterialModel::MaterialModelInputs<dim> in(quadrature.size(),
+                                                 this->n_compositional_fields());
+
+      unsigned i = 0;
+      for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+        {
+          if (cell->is_locally_owned())
+            {
+              bool coarsen = false;
+              bool clear_refine = false;
+              bool refine = false;
+              bool clear_coarsen = false;
+
+              fe_values.reinit(cell);
+              in.reinit(fe_values, cell, this->introspection(), this->get_solution(), true);
+
+              for (unsigned int i_quad=0; i_quad<quadrature.size(); ++i_quad)
+                {
+                  for (auto &isosurface : isosurfaces)
+                    {
+                      // setup the vector to check
+                      std::vector<double> values(isosurface.min_values.size());
+                      for (unsigned int index = 0; index < isosurface.properties.size(); ++index)
+                        {
+                          if (isosurface.properties[index].name == Internal::PropertyName::Temperature)
+                            {
+                              values[index] = in.temperature[i_quad];
+                            }
+                          else if (isosurface.properties[index].name == Internal::PropertyName::Composition)
+                            {
+                              values[index] = std::min(std::max(in.composition[i_quad][isosurface.properties[index].index], 0.0), 1.0);
+                            }
+                          else if (isosurface.properties[index].name == Internal::PropertyName::Background)
+                            {
+                              double sum_composition = 0.0;  // Compute background material fraction
+                              for (unsigned i=0 ; i < in.composition[i_quad].size() ; ++i)
+                                {
+                                  sum_composition += std::min(std::max(in.composition[i_quad][i], 0.0), 1.0);
```

Comment:
You assume that all compositional fields participate in a case where they sum up to 1-background. That is quite restrictive as compositional fields are also used for other things.

Reply:
same as comment above, @lhy11009 what is your opinion on this?